gquant/dataframe_flow/node.py

```diff
@@ -284,44 +562,193 @@ def load_cache(self, filename):
+                outputs_dly.setdefault(oport, []).append(oport_out.persist())
+
+        # DEBUGGING
+        # print('OUTPUTS_DLY:\n{}'.format(outputs_dly))
+
+        output_df = {}
+        # A dask_cudf object is  synthesized from a list of delayed objects.
+        # Per outputs_dly above use dask_cudf.from_delayed API.
+        for oport in self.get_output_ports():
+            output_df[oport] = dask_cudf.from_delayed(outputs_dly[oport])
+
+        return output_df
+
+    def __call__(self, inputs_data):
         Class = type(self)
         cache = Class.cache_dir
-        inputs = [self.__make_copy(i) for i in inputs]
+
+        if self.using_ports():
+            # nodes with ports take dictionary as inputs
+            inputs = {iport: self.__make_copy(data_input)
+                      for iport, data_input in inputs_data.items()}
+        else:
+            # nodes without ports take list as inputs
+            inputs = [self.__make_copy(data_input)
+                      for data_input in inputs_data.values()]
+
         if not isinstance(self.load, bool) or self.load:
             if isinstance(self.load, bool):
-                output_df = self.load_cache(cache+'/'+self.uid+'.hdf5')
+                # TODO: Load CACHE support for ports based API.
+                output_df = self.load_cache(cache + '/' + self.uid + '.hdf5')
             else:
                 output_df = self.load
         else:
             if not self.delayed_process:
                 output_df = self.process(inputs)
             else:
-                # handle the dask dataframe automatically
-                # use the to_delayed interface
-                # TODO, currently only handles first input is dask_cudf df
-                i_df = inputs[0]
-                rest = inputs[1:]
-                if isinstance(i_df,  dask_cudf.DataFrame):
-                    d_fun = dask.delayed(self.process)
-                    output_df = dask_cudf.from_delayed([
-                        d_fun([item] + rest) for item in i_df.to_delayed()])
+                if self.using_ports():
+                    use_delayed = self.__check_dly_processing_prereq(inputs)
+                    if use_delayed:
+                        output_df = self.__delayed_call(inputs)
+                    else:
+                        output_df = self.process(inputs)
                 else:
-                    output_df = self.process(inputs)
+                    # handle the dask dataframe automatically
+                    # use the to_delayed interface
+                    # TODO, currently only handles first input is dask_cudf df
+                    i_df = inputs[0]
+                    rest = inputs[1:]
+                    if isinstance(i_df, dask_cudf.DataFrame):
+                        d_fun = dask.delayed(self.process)
+                        output_df = dask_cudf.from_delayed([
+                            d_fun([item] + rest)
+                            for item in i_df.to_delayed()])
```

Comment:
If `persist` method is required to avoid redundancy, should I call the `persist` method here? 
I am hacking a solution here that it only handles the case that the first input is dask_cudf. We should move away from this old method to use the new input ports approach.

Reply:
I think it's fine because it's all done in one call using `from_delayed`. They way it's written it will not handle multiple dask_cudf dataframes though. As long as only the first and only one input is a dask-cudf dataframe, other inputs can be anything but not dask-cudf dataframes, it will work.

In the ports case I need an additional delayed function `get_pout` that confused the dask-scheduler.

Reply:
okay